synth_waveform_F32.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * AudioSynthWaveform_F32
+ * 
+ * Created: Patrick Radius, January 2017
+ * Purpose: Generate waveforms at a given frequency and amplitude. Allows for pitch-modulation and portamento.
+ *          
+ * This processes a single stream fo audio data (ie, it is mono)       
+ *          
+ * MIT License.  use at your own risk.
+*/
+#ifndef SYNTHWAVEFORMF32_H
+#define SYNTHWAVEFORMF32_H
+
+#include <arm_math.h>
+#include <AudioStream_F32.h>
+
+class AudioSynthWaveform_F32 : public AudioStream_F32
+{
+  public:
+    enum OscillatorMode {
+        OSCILLATOR_MODE_SINE = 0,
+        OSCILLATOR_MODE_SAW,
+        OSCILLATOR_MODE_SQUARE,
+        OSCILLATOR_MODE_TRIANGLE
+    };
+
+    AudioSynthWaveform_F32(void) : AudioStream_F32(1, inputQueueArray_f32),
+                _PI(2*acos(0.0f)),
+                twoPI(2 * _PI),
+                _OscillatorMode(OSCILLATOR_MODE_SINE),
+                _Frequency(440.0f),
+                _Phase(0.0f),
+                _PhaseIncrement(0.0f),
+                _PitchModAmt(0.0f),
+                _PortamentoIncrement(0.0f),
+                _PortamentoSamples(0),
+                _CurrentPortamentoSample(0),
+                _NotesPlaying(0) {};
+
+    void frequency(float32_t freq) {
+        float32_t nyquist = AUDIO_SAMPLE_RATE_EXACT/2;
+
+        if (freq < 0.0) freq = 0.0;
+        else if (freq > nyquist) freq = nyquist;
+
+        if (_PortamentoSamples > 0 && _NotesPlaying > 0) {
+          _PortamentoIncrement = (freq - _Frequency) / (float32_t)_PortamentoSamples;
```

Comment:
Also, what happens if a second frequency command occurs while the portamento is still sliding the pitch from the first command?  Will the pitch start and end at the right place?

For "start", I think that the new slide should start at the pitch that we're at currently, even if it's an intermediate pitch due to portamento on the first note.

For "end", it should end (obviously) on the pitch that the user expects.

This is your code, so you get to choose the portamento behavior.  I'm just asking if you've considered the possibilities of what might happen, especially if a long portamento is asked for.